[ACS][CallAutomatoin] Easily Handling EventHandler

minwoolee-msft opened the pull request:
ACS Call Automation is looking to add a new Event Processor to provide an easier way of handling event.

**Allow Event Processor to process event**
```csharp
// wait for event...
_eventHandler = _callAutomationClient.GetEventProcessor();
_eventHandler.ProcessEvents(cloudEvents);
```

**Waiting on Event with result model (easiest way)**
```csharp
AddParticipantsResult result= callConnection.AddParticipants(...);
AddParticipantsEventResult eventResult = await result.WaitForEvent(); // will wait until event arrives, or times out
AddParticipantsSucceeded myEvent = eventResult.SuccessEvent; // This is AddParticipantSucceeded event!
```

**Waiting on any event with predicate (more advanced)**
```csharp
// will wait until event arrives, or times out for Call disconnect
CallDisconnected disconnectedEvent = (CallDisconnected)await _eventHandler.WaitForEvent(predicate 
=> predicate.CallConnectionId == myConnectionId 
&& predicate.GetType() == typeof(CallDisconnected)); 
```

**Waiting any event any time, then run delegates on match (great for non-sequential events, like ParticipantsUpdated or CallDisconnect)**
```csharp
// whenever participantsUpdated event comes, print out # of people in the call!
_eventHandler.AttachOngoingEventProcessor<ParticipantsUpdated>(myCallConnectionId,
e => _logger.LogInformation($"Current # of People in the call: {e.Participants.Count}"));
```

**Significant changes that need attention:**
1. All operation generate operation context if not provided, SDK will generate one for them and send in the request. This will not impact any existing behaviors- but this is required to have eventProcessor work properly.
2. Some methods that did not return result, will now return Result instead of Void (such as, PlayResult). This is required for it to work with Result.WaitForEvent() Pattern. This will not impact any existing behaviors.

minwoolee-msft committed:
Initial commit to the new event handler

minwoolee-msft committed:
- Adding options for EventHandler
- AddOrUpdate on eventsbacklog
- Accepting Generic Type as input for ease of use
- Summaries updated
- Adding more test

minwoolee-msft committed:
- delete all memory items when call disconnection arrives
- adding more test

azure-sdk commented:
**API change check**

APIView has identified API level changes in this PR and created following API reviews.

[Azure.Communication.CallAutomation](https://apiview.dev/Assemblies/Review/616d083d0b1340f2a713115b478c3ebf)

minwoolee-msft committed:
- Adding more tests
- Updated api

minwoolee-msft committed:
- Fix with removing events from list

DominikMe commented on sdk/communication/Azure.Communication.CallAutomation/src/EventHandler/EventAwaiterOngoing.cs:
```diff
@@ -0,0 +1,29 @@
+        private string _callConnectionId;
+        private Action<TEvent> _eventHandler;
+
+        internal Action<object, CallAutomationEventArgs> OnEventRecieved => OnEventsReceived;
```
typo:
```suggestion
        internal Action<object, CallAutomationEventArgs> OnEventReceived => OnEventsReceived;
```
(and update references)

DominikMe commented on sdk/communication/Azure.Communication.CallAutomation/src/EventHandler/CallAutomationEventHandler.cs:
```diff
@@ -0,0 +1,251 @@
+        /// <param name="events">Incoming CallAutomationEventBase object.</param>
+        public void ProcessEvents(IEnumerable<CallAutomationEventBase> events)
+        {
+            CallAutomationEventBase recievedEvent = events.FirstOrDefault();
```
typo
```suggestion
            CallAutomationEventBase receivedEvent = events.FirstOrDefault();
```

minwoolee-msft committed:
- typeo fix
- upper limit on eventbacklogs
- constant value for timeout seconds & limits

petrsvihlik commented on sdk/communication/Azure.Communication.CallAutomation/src/EventHandler/EventBacklog.cs:
```diff
@@ -0,0 +1,89 @@
+            }
+        }
+
+        internal bool GetAndRemoveEvent(IEnumerable<Type> eventTypes, string callConnectionId, string operationContext, out KeyValuePair<string, CallAutomationEventBase> matchingEvent)
```
This is a typical example for Try Pattern. The method should be prefixed with `Try`

petrsvihlik commented on sdk/communication/Azure.Communication.CallAutomation/src/EventHandler/EventBacklog.cs:
```diff
@@ -0,0 +1,89 @@
+        /// <param name="backlogEventId">Internally used id for tracking the saved event.</param>
+        /// <param name="eventsToBeSaved">Incoming Event to be saved.</param>
+        /// <returns></returns>
+        internal bool AddEvent(string backlogEventId, CallAutomationEventBase eventsToBeSaved)
```
This is a typical example for Try Pattern. The method should be prefixed with `Try`

petrsvihlik commented on sdk/communication/Azure.Communication.CallAutomation/src/EventHandler/EventBacklog.cs:
```diff
@@ -0,0 +1,89 @@
+        /// </summary>
+        /// <param name="internalEventId">Key of the event in events backlog.</param>
+        /// <returns></returns>
+        internal bool RemoveEvent(string internalEventId) => _eventBacklog.TryRemove(internalEventId, out _);
```
TryRemoveEvent

minwoolee-msft commented on sdk/communication/Azure.Communication.CallAutomation/src/EventHandler/CallAutomationEventHandler.cs:
```diff
@@ -0,0 +1,251 @@
+        /// <param name="events">Incoming CallAutomationEventBase object.</param>
+        public void ProcessEvents(IEnumerable<CallAutomationEventBase> events)
+        {
+            CallAutomationEventBase receivedEvent = events.FirstOrDefault();
```
The reason is that events received will always be (as of now) single event.
I think when I was doing development, I assumed there will be no changes that will register more than 1 (as that is the case right now)
I can update this to work with multiple events passed, for future proof.

petrsvihlik replied:
thanks for explaining!

>I can update this to work with multiple events passed, for future proof.

either that or at least leave a comment for future generations reading the code :) it's not intuitive at the first glance.

minwoolee-msft commented on sdk/communication/Azure.Communication.CallAutomation/src/EventHandler/CallAutomationEventArgs.cs:
```diff
@@ -0,0 +1,14 @@
+{
+    internal class CallAutomationEventArgs : EventArgs
+    {
+        internal string eventArgsId { get; set; }
+
+        internal CallAutomationEventBase callAutomationEvent { get; set; }
```
Thanks - Will Update.

minwoolee-msft commented on sdk/communication/Azure.Communication.CallAutomation/src/EventHandler/EventAwaiter.cs:
```diff
@@ -0,0 +1,69 @@
+            taskSource.TrySetException(new[] { new TimeoutException() });
+        }
+
+        public void Dispose()
+        {
+            _expiringTimer.Dispose();
+        }
```
Thank you - will read and follow the guideline's suggestion.

minwoolee-msft committed:
- Updated Name to EventProcessor, as suggested
- Added results for methods that is missing Result output
- All results now have WaitForEvent()
- Results' WaitForEvent() waits for the method's specific EventResult. Success/Failure events can be retrieved accordingly
- Updated Dispose()
- Added tests for AttachEventProcessor
- Added tests for ResultWithEventProcessor
- Updated naming on variables and methods, as suggested
- Updated spelling/grammar mistakes
- Various bug fixes

minwoolee-msft commented on sdk/communication/Azure.Communication.CallAutomation/src/EventHandler/CallAutomationEventHandler.cs:
```diff
@@ -0,0 +1,251 @@
+    /// <summary>
+    /// Call Automation's EventHandler for incoming events for ease of use.
+    /// </summary>
+    public class CallAutomationEventHandler
```
Updated to EventProcessor

minwoolee-msft commented on sdk/communication/Azure.Communication.CallAutomation/src/EventHandler/CallAutomationEventHandler.cs:
```diff
@@ -0,0 +1,251 @@
+            where T3 : CallAutomationEventBase
+            where T4 : CallAutomationEventBase
+            where T5 : CallAutomationEventBase
+            => await WaitForEvent(new List<Type> { typeof(T1), typeof(T2), typeof(T3), typeof(T4), typeof(T5) }, callConnectionId, operationContext).ConfigureAwait(false);
```
Added a lot more testing for each cases

minwoolee-msft committed:
Merge branch 'main' into minwoolee/eventhandler2

minwoolee-msft committed:
- Bug fix on RemoveParticipants operation context generation

minwoolee-msft committed:
- Adding custom timeouts for WaitForEvent

minwoolee-msft committed:
- APIfile update

minwoolee-msft committed:
Fixed bug with Media's operationcontext generation

minwoolee-msft committed:
- allow no operation context to match any event

petrsvihlik commented on sdk/communication/Azure.Communication.CallAutomation/src/EventProcessor/EventProcessor.cs:
```diff
@@ -0,0 +1,176 @@
+        /// <param name="predicate">Predicate for waiting on event</param>
+        /// <param name="eventTimeout">Timeout for this waiting on event</param>
+        /// <returns>Returns CallAutomationEvent once matching event arrives.</returns>
+        public async Task<CallAutomationEventBase> WaitForEvent(Func<CallAutomationEventBase, bool> predicate, TimeSpan eventTimeout = default)
```
do we want/need to expose it publicly? are we expecting customers to come up with more complex/granular predicates than those defined by us?

if so, maybe it'd make sense to distinguish the method even further from Attach/Detach methods by naming it `WaitForSingleEvent`. 

and/or provide some overload(s) for common operations... like `WaitForSingleEvent<TEvent>(string connectionId = null, string operationContext = null, Func<CallAutomationEventBase, bool> predicate = null)`

minwoolee-msft replied:
Thank you for the suggestion - I agree that waiting for single event with old way of doing things (with generic event <TEvent>) is very easy to use, and I can see this is 90% of the time how devs would use it. I will have them added as overload.

petrsvihlik reviewed (commented):
mostly nits

minwoolee-msft commented on sdk/communication/Azure.Communication.CallAutomation/src/EventProcessor/EventResult/StartRecognizingEventResult.cs:
```diff
@@ -0,0 +1,28 @@
+namespace Azure.Communication.CallAutomation
+{
+    /// <summary>
+    /// <summary> StartRecognizingEventResult is returned from WaitForEvent of StartRecognizingResult. </summary>
```
Thanks - will update all of them.

minwoolee-msft commented on sdk/communication/Azure.Communication.CallAutomation/src/EventProcessor/EventAwaiter.cs:
```diff
@@ -0,0 +1,79 @@
+            // see if events sent matches filter set on constructor.
+            if (_predicate(e.CallAutomationEvent))
+            {
+                // dispose expiring timer, as we don't want timer activating.
```
Yep - will go through the items

minwoolee-msft committed:
- Updated comments and summaries
- Naming Update
- Added new overload for WaitForSingleEvent

minwoolee-msft committed:
More spelling and comments updates

minwoolee-msft committed:
Merge branch 'main' into minwoolee/eventhandler2
